internal/sbi/httpcallback/api_handle_dereg_notification.go

```diff
@@ -0,0 +1,105 @@
+	"github.com/free5gc/amf/internal/sbi/consumer"
+	"github.com/free5gc/openapi"
+	"github.com/free5gc/openapi/models"
+)
+
+func HTTPAmfHandleDeregistrationNotification(c *gin.Context) {
+	// TS 23.502 - 4.2.2.2.2 - step 14d
+	logger.CallbackLog.Infoln("Handle Deregistration Notification")
+
+	var deregData models.DeregistrationData
+
+	requestBody, err := c.GetRawData()
+	if err != nil {
+		logger.CallbackLog.Errorf("Get Request Body error: %+v", err)
+		problemDetails := models.ProblemDetails{
+			Title:  "System failure",
+			Status: http.StatusInternalServerError,
+			Detail: err.Error(),
+			Cause:  "SYSTEM_FAILURE",
+		}
+		c.JSON(http.StatusInternalServerError, problemDetails)
+		return
+	}
+
+	err = openapi.Deserialize(&deregData, requestBody, "application/json")
+	if err != nil {
+		problemDetails := "[Request Body] " + err.Error()
+		rsp := models.ProblemDetails{
+			Title:  "Malformed request syntax",
+			Status: http.StatusBadRequest,
+			Detail: problemDetails,
+		}
+		logger.CallbackLog.Errorln(problemDetails)
+		c.JSON(http.StatusBadRequest, rsp)
+		return
+	}
+
+	ueid := c.Param("ueid")
+	amfSelf := amf_context.GetSelf()
+	ue, ok := amfSelf.AmfUeFindByUeContextID(ueid)
+	if !ok {
+		logger.CallbackLog.Errorf("AmfUe Context[%s] not found", ueid)
+		problemDetails := &models.ProblemDetails{
+			Status: http.StatusNotFound,
+			Cause:  "CONTEXT_NOT_FOUND",
+		}
+		c.JSON(http.StatusNotFound, problemDetails)
+		return
+	}
+
+	switch deregData.DeregReason {
+	case models.DeregistrationReason_UE_INITIAL_REGISTRATION:
+		// TS 23.502 - 4.2.2.2.2 General Registration
+		// Invokes the Nsmf_PDUSession_ReleaseSMContext for the corresponding access type
+		ue.SmContextList.Range(func(key, value interface{}) bool {
+			smContext := value.(*amf_context.SmContext)
+
+			if smContext.AccessType() == deregData.AccessType {
+				var problemDetails *models.ProblemDetails
+				problemDetails, err = consumer.SendReleaseSmContextRequest(ue, smContext, nil, "", nil)
+				if problemDetails != nil {
+					ue.GmmLog.Errorf("Release SmContext Failed Problem[%+v]", problemDetails)
+				} else if err != nil {
+					ue.GmmLog.Errorf("Release SmContext Error[%v]", err.Error())
+				}
+			}
+
+			return true
+		})
+	case models.DeregistrationReason_SUBSCRIPTION_WITHDRAWN:
+		// TS 23.502 - 4.2.2.3.3 Network-initiated Deregistration
+		// The AMF executes Deregistration procedure over the access(es) the Access Type indicates
+
+		// Use old AMF as the backup AMF
+		backupAmfInfo := models.BackupAmfInfo{
+			BackupAmf: amfSelf.Name,
+			GuamiList: amfSelf.ServedGuamiList,
+		}
+		ue.UpdateBackupAmfInfo(backupAmfInfo)
+	}
+
+	// TS 23.502 - 4.2.2.2.2 General Registration
+	// The old AMF should clean the UE context
+	// The AMF also unsubscribes with the UDM using Nudm_SDM_Unsubscribe service operation.
+	err = gmm_common.PurgeSubscriberData(ue, deregData.AccessType)
+	if err != nil {
+		logger.CallbackLog.Errorf("Purge subscriber data Error: %v", err)
+	}
```

Comment:
Original RemoveAmfUe at dereg when notifyNf is true also considers AMPolicyControlDelete. Should AMPolicyControlDelete be included here when the cause is not UE_REGISTRATION_AREA_CHANGE? (According to TS 23.502 4.2.2.2.2 14d, the new AMF (cause: UE_REGISTRATION_AREA_CHANGE) will reuse PCF ID.)

Reply:
I don't find the description in R15 - TS 23.502 - 4.2.2.2.2 - 14d and I am not sure if we need to delete AmPolicy here.
Based on my understanding of implicitly deregistering AMF, the goal of it is only to remove the UE context from old AMF rather than deregister the UE from the network.

Reply:
Sorry, I think that's step 20. in R15.
> 20. [Conditional] old AMF to (V-)PCF: AMF-Initiated Policy Association Termination. 
 If the old AMF previously initiated a Policy Association to the PCF, and the old AMF did not transfer the PCF 
ID(s) to the new AMF (e.g. new AMF is in different PLMN), the old AMF performs an AMF-initiated Policy 
Association Termination procedure, as defined in clause 4.16.3.2, to delete the association with the PCF.

Reply:
For UE_INITIAL_REGISTRATION and SUBSCRIPTION_WITHDRAW, it is fine to do AMF-Initiated Policy Association Termination directly. However, if it is REGISTRATION_AREA_CHANGE, it is possible that the new AMF reuses the associated PCF ID, which hence should not be terminated.
R15 spec mentioned the condition: 
> If the old AMF previously initiated a Policy Association to the PCF, and the old AMF did not transfer the PCF
ID(s) to the new AMF

R16 has clearer instructions:
> If the old AMF has established an AM Policy Association and a UE Policy Association with the PCF(s), and the 
old AMF did not transfer the PCF ID(s) to the new AMF (e.g. new AMF is in different PLMN), the old AMF 
performs an AMF-initiated Policy Association Termination procedure, as defined in clause 4.16.3.2, and 
performs an AMF-initiated UE Policy Association Termination procedure, as defined in clause 4.16.13.1. In 
addition, if the old AMF transferred the PCF ID(s) in the UE context but the new AMF informed in step 10 that 
the AM Policy Association information and UE Policy Association information in the UE context will not be 
used then the old AMF performs an AMF-initiated Policy Association Termination procedure, as defined in 
clause 4.16.3.2, and performs an AMF-initiated UE Policy Association Termination procedure, as defined in 
clause 4.16.13.1.

Reply:
> For UE_INITIAL_REGISTRATION and SUBSCRIPTION_WITHDRAW, it is fine to do AMF-Initiated Policy Association Termination directly. However, if it is REGISTRATION_AREA_CHANGE, it is possible that the new AMF reuses the associated PCF ID, which hence should not be terminated.

I agree with that, however, free5GC currently doesn't utilize the PCF ID to select the PCF, therefore, I leave the TODO comments for it.
Should we complete the feature (searching PCF by using PCF ID) in this PR or open another PR for this feature?

Reply:
I think it is fine to leave for a new PR. Lastly, please help to change the TODO comments (`// TODO: It also needs to check if the PCF ID is tranfered to new AMF...`) to the followings:
```
// For UE_INITIAL_REGISTRATION and SUBSCRIPTION_WITHDRAW, do AMF-Initiated Policy Association Termination directly.
// TODO: For REGISTRATION_AREA_CHANGE, old AMF performs an AMF-initiated Policy Association Termination 
// procedure if the old AMF has established an AM Policy Association and a UE Policy Association with the PCF(s), 
// and the old AMF did not transfer the PCF ID(s) to the new AMF. (Ref: TS 23.502 - 4.2.2.2.2) 
```

Reply:
OK